config/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurerTests.java

```diff
@@ -526,6 +528,107 @@ public void requestWhenSessionManagementConfiguredThenUserConfigurationOverrides
 		assertThat(result.getRequest().getSession(false)).isNotNull();
 	}
 
+	// -- custom bearer token resolver
+
+	@Test
+	public void requestWhenBearerTokenResolverAllowsRequestBodyThenEitherHeaderOrRequestBodyIsAccepted()
+			throws Exception {
+
+		this.spring.register(AllowBearerTokenInRequestBodyConfig.class, JwtDecoderConfig.class,
+				BasicController.class).autowire();
+
+		JwtDecoder decoder = this.spring.getContext().getBean(JwtDecoder.class);
+		when(decoder.decode(anyString())).thenReturn(JWT);
+
+		this.mvc.perform(get("/authenticated")
+				.with(bearerToken(JWT_TOKEN)))
+				.andExpect(status().isOk())
+				.andExpect(content().string(JWT_SUBJECT));
+
+		this.mvc.perform(post("/authenticated")
+				.param("access_token", JWT_TOKEN))
+				.andExpect(status().isOk())
+				.andExpect(content().string(JWT_SUBJECT));
+	}
+
+	@Test
+	public void requestWhenBearerTokenResolverAllowsQueryParameterThenEitherHeaderOrQueryParameterIsAccepted()
+			throws Exception {
+
+		this.spring.register(AllowBearerTokenAsQueryParameterConfig.class, JwtDecoderConfig.class,
+				BasicController.class).autowire();
+
+		JwtDecoder decoder = this.spring.getContext().getBean(JwtDecoder.class);
+		when(decoder.decode(anyString())).thenReturn(JWT);
+
+		this.mvc.perform(get("/authenticated")
+				.with(bearerToken(JWT_TOKEN)))
+				.andExpect(status().isOk())
+				.andExpect(content().string(JWT_SUBJECT));
+
+		this.mvc.perform(get("/authenticated")
+				.param("access_token", JWT_TOKEN))
+				.andExpect(status().isOk())
+				.andExpect(content().string(JWT_SUBJECT));
+	}
+
+	@Test
+	public void getBearerTokenResolverWhenDuplicateResolverBeansAndAnotherOnTheDslThenTheDslOneIsUsed() {
+		BearerTokenResolver resolverBean = mock(BearerTokenResolver.class);
+		BearerTokenResolver resolver = mock(BearerTokenResolver.class);
+
+		GenericWebApplicationContext context = new GenericWebApplicationContext();
+		context.registerBean("resolverOne", BearerTokenResolver.class, () -> resolverBean);
+		context.registerBean("resolverTwo", BearerTokenResolver.class, () -> resolverBean);
+		this.spring.context(context).autowire();
+
+		OAuth2ResourceServerConfigurer oauth2 = new OAuth2ResourceServerConfigurer(context);
+
+		oauth2.bearerTokenResolver(resolver);
+
+		assertThat(oauth2.getBearerTokenResolver()).isEqualTo(resolver);
```

Comment:
I see why you made `getBearerTokenResolver()` package private. However, I'd rather avoid opening up methods so it can work for testing.

This can be tested by issuing a request and then `verify()` or `verifyZeroInteractions()` to assert which resolver is in effect.

Reply:
Can you elaborate on how this confirms the intended behavior? That is, we want to confirm that the correct implementation is picked. We won't actually invoke the mock without performing some kind of mock mvc roundtrip, which, this being a unit test, might be a bit odd.

We did the same thing with `getJwtDecoder` in `JwtConfigurer` to achieve what appears to me to be following the same reasoning, e.g. corner case should be a unit test, so we make the method package-private so the unit test can confirm. I agree that bumping up the access level of a method for a unit test is worth thinking twice about, but I don't see another way to do it other than reflection or turning it into an integration test.

Now, we could look at this becoming an integration test; however, that seems to deviate from the original thrust of this set of tests. (Again, for context, take a look at the tests for JwtConfigurer and see that some are integration tests and some are unit tests.)

Reply:
This class mixes unit tests and integration tests which we typically separate in different classes. Also, `SecurityConfigurer`s are integration testing. Are there other configuers that do unit testing with the exception of `JwtConfigurer`?